tethysext/atcore/tests/integrated_tests/models/files_database/file_collection_client_tests.py

```diff
@@ -463,3 +474,327 @@ def test_bad_collection_client(self):
+            file_text = f.write('This text should be written to file.')
+        with open(os.path.join(collection_client.path, 'file1.txt')) as of:
+            file_text = of.read()
+            self.assertEqual('This text should be written to file.', file_text)
+
+    def test_open_file_does_not_exist(self):
+        """Test opening a file that doesn't exist throws the correct error."""
+        test_dir_name = 'test_open_file_does_not_exist'
+        base_files_root_dir = os.path.join(self.test_files_base, test_dir_name)
+        root_dir = os.path.join(self.test_files_base, 'temp', test_dir_name)
+        self.copy_files_to_temp_directory(base_files_root_dir, root_dir)
+        _, collection_instance = self.get_database_and_collection(
+            database_id=self.general_database_id, collection_id=self.general_collection_id,
+            root_directory=root_dir, database_meta={}, collection_meta={}
+        )
+        collection_client = FileCollectionClient(self.session, self.general_collection_id)
+        with self.assertRaises(FileCollectionItemNotFoundError) as exc:
+            with collection_client.open_file('file1.txt', 'w') as _:
+                pass
+        self.assertTrue('"file1.txt" not found in this collection.' in str(exc.exception))
+
+    def test_open_file_directory(self):
+        """Test opening a directory throws an error."""
+        test_dir_name = 'test_open_file_directory'
+        base_files_root_dir = os.path.join(self.test_files_base, test_dir_name)
+        root_dir = os.path.join(self.test_files_base, 'temp', test_dir_name)
+        self.copy_files_to_temp_directory(base_files_root_dir, root_dir)
+        _, collection_instance = self.get_database_and_collection(
+            database_id=self.general_database_id, collection_id=self.general_collection_id,
+            root_directory=root_dir, database_meta={}, collection_meta={}
+        )
+        collection_client = FileCollectionClient(self.session, self.general_collection_id)
+        with self.assertRaises(IsADirectoryError):
+            with collection_client.open_file('dir1', 'r') as _:
+                pass
+
+    def test_walk(self):
+        """Test walking a collection."""
+        test_dir_name = 'test_walk'
+        base_files_root_dir = os.path.join(self.test_files_base, test_dir_name)
+        root_dir = os.path.join(self.test_files_base, 'temp', test_dir_name)
+        self.copy_files_to_temp_directory(base_files_root_dir, root_dir)
+        _, collection_instance = self.get_database_and_collection(
+            database_id=self.general_database_id, collection_id=self.general_collection_id,
+            root_directory=root_dir, database_meta={}, collection_meta={}
+        )
+        collection_client = FileCollectionClient(self.session, self.general_collection_id)
+        walk_output = [x for x in collection_client.walk()]
+        expected_walk_output = [
+            ('.', ['dir1'], ['file1.txt']),
```

Comment:
Now that I'm thinking of it, I think "." should change once you are past the first level of directories... I think there is a built-in function that will convert one path to a relative path given some reference directory. That's probably what we want to use.

Reply:
done.

Reply:
The way you did it probably works. (not sure what happens if you join with a path with a  dot in it). It would probably be saver to use the built-in function `os.path.relpath` with `file_collection_client.path` set as the `start`.
See: https://docs.python.org/3/library/os.path.html#os.path.relpath

Add another test in which we call walk() and open each file using .open_file(). That will require doing a join with the base path.

Reply:
this is done.